File: easy/SymmetricTree.ts

// Definition for a binary tree node.
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

function isSymmetric(root: TreeNode | null): boolean {
    if (root === null) return true
    // Handle special case
    if (root.left !== null && root.right!== null && root.left.val !== root.right.val) return false
    let valArray = inOrderTravere(root)
    if (valArray.length % 2 == 0) return false
    return valArray.toString() === valArray.reverse().toString()
};

// Traverse order: left, root, right
function inOrderTravere(root: TreeNode | null): number[] {
    // Handle special case
    if (root == null) return [-999]
    if (root.left === null && root.right == null) return [root.val]
    return inOrderTravere(root.left).concat([root.val]).concat(inOrderTravere(root.right))
}

// Special cases:
// [1,2,2,2,null,2]
// [5,4,1,null,1,null,4,2,null,2,null]
// Notice: the array is formed from the tree top-down layer by layer

const node7 = null;
const node6 = new TreeNode(2);
const node5 = null;
const node4 = new TreeNode(2);
const node3 = new TreeNode(2, node6, node7);
const node2 = new TreeNode(2, node4, node5);
const root = new TreeNode(1, node2, node3);

console.log(isSymmetric(root))

File: easy/symmetric_tree.py
# Definition for a binary tree node.
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None: return True
        return self.check(root.left, root.right)
    
    # check the equility of left and right simultaneously in each recursion 
    def check(self, left, right):
        if left is None and right is None:
            return True
        if (left is None and right is not None) or (left is not None and right is None):
            return False
        if left.val != right.val:
            return False
        outside = self.check(left.left, right.right)
        inside = self.check(left.right, right.left)
        return outside and inside


root = TreeNode(1, TreeNode(3, TreeNode(4, None, None), TreeNode(5, TreeNode(8, None, None), TreeNode(9, None, None))), TreeNode(3, TreeNode(5, TreeNode(9, None, None), TreeNode(8, None, None)), TreeNode(4, None, None)))

solver = Solution()
print(solver.isSymmetric(root))
    